File: src/main/java/at/searles/parsing/combinators/ReducerRep.java
package at.searles.parsing.combinators;

import at.searles.parsing.*;
import at.searles.parsing.printing.ConcreteSyntaxTree;
import at.searles.parsing.printing.PartialConcreteSyntaxTree;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Collections;

/**
 * Parser for repetitions
 */
public class ReducerRep<T> implements Reducer<T, T>, Recognizable.Rep {
    private final Reducer<T, T> parent;

    public ReducerRep(Reducer<T, T> parent) {
        this.parent = parent;
    }

    @Override
    public T parse(ParserStream stream, @NotNull T left) {
        long preStart = stream.getStart();

        while (true) {
            T t = parent.parse(stream, left);

            // Contract of Reducer
            assert stream.getStart() == preStart;

            if (t == null) break;

            left = t;
        }

        return left;
    }


    @Override
    public Recognizable parent() {
        return parent;
    }

    @Override
    public PartialConcreteSyntaxTree<T> print(@NotNull T t) {
        T left = t;

        ArrayList<ConcreteSyntaxTree> trees = new ArrayList<>();

        for (; ; ) {
            PartialConcreteSyntaxTree<T> next = parent.print(left);

            if (next == null) {
                break;
            }

            trees.add(next.right);

            left = next.left;
        }

        Collections.reverse(trees);

        return new PartialConcreteSyntaxTree<>(left, ConcreteSyntaxTree.fromList(trees));
    }

    @Override
    public String toString() {
        return createString();
    }
}

File: src/main/java/at/searles/parsing/combinators/ReducerRep.kt
package at.searles.parsing.combinators

import at.searles.parsing.ParserStream
import at.searles.parsing.Recognizable.Rep
import at.searles.parsing.Reducer
import at.searles.parsing.printing.ConcreteSyntaxTree
import at.searles.parsing.printing.PartialConcreteSyntaxTree
import java.util.*

/**
 * Parser for repetitions
 */
class ReducerRep<T>(override val parent: Reducer<T, T>) : Reducer<T, T>, Rep {

    override fun parse(stream: ParserStream, input: T): T {

        var item = input
        val preStart = stream.start

        while (true) {
            val nextItem = parent.parse(stream, item)
            assert(stream.start == preStart)
            if (nextItem == null) return item
            item = nextItem
        }
    }

    override fun print(item: T): PartialConcreteSyntaxTree<T>? {
        var left: T = item
        val trees = ArrayList<ConcreteSyntaxTree>()

        while (true) {
            val next = parent.print(left) ?: break
            trees.add(next.right)
            left = next.left
        }

        trees.reverse()

        return PartialConcreteSyntaxTree(left, ConcreteSyntaxTree.fromList(trees))
    }

    override fun toString(): String {
        return createString()
    }
}